File: NB_sentiment_analysis_IMDB.R
library(tm)
library(wordcloud)
library(RColorBrewer)
library(dplyr)

setwd("~/Documents/Machine Learning/ML_project")

## Stage 1 ##

# Read in the training and test (cleaned) documents

train_df <- read.csv("full_train_sentiment.csv" , stringsAsFactors = F)
train_df$sentiment <- factor(train_df$sentiment)

test_df <- read.csv("sentiment_test.csv" , stringsAsFactors = F)
test_df$sentiment <- factor(test_df$sentiment)

# Create a corpus of training documents in order to carry out text processing

train_corpus <- Corpus(VectorSource(train_df$review))

train_corpus <- tm_map(train_corpus , tolower)
train_corpus <- tm_map(train_corpus , removeNumbers)
train_corpus <- tm_map(train_corpus , removePunctuation)
train_corpus <- tm_map(train_corpus , removeWords , stopwords(kind = "english"))
train_corpus <- tm_map(train_corpus , removeWords , c("one" , "two" , "movie" , "movies", "film" , "films" , "get" , "make"))
train_corpus <- tm_map(train_corpus , stripWhitespace)
train_corpus <- tm_map(train_corpus , stemDocument , language = "english")

# Create a document-term matrix
# Notice how large the dimensions are (2001 rows of 30550 attributes)

train_dtm <- DocumentTermMatrix(train_corpus)
dim(train_dtm)

# Remove sparse terms to reduce this high dimensionality
# Dimensions reduced to 2001 rows of 195 attributes

train_dtm <- removeSparseTerms(train_dtm , 0.80)
dim(train_dtm)

# Which words occur most frequently in the training documents?

freq_word <- colSums(as.matrix(train_dtm))

# Wordcloud of most frequent words (found at least 1000 times)

wordcloud(names(freq_word) , freq_word , min.freq = 1000 , colors = brewer.pal(7 , "Dark2") , scale = c(2 , 1))

# The document term matrix can be coerced to a dataframe and we can attach the sentiment labels

full_freq_df <- data.frame(as.matrix(train_dtm))
full_freq_df <- cbind(full_freq_df , sentiment = train_df$sentiment)

# We need apriori probabilities for each sentiment level

apriori_pos <- nrow(train_df %>%
                      filter(sentiment == "positive")) / nrow(train_df)

apriori_neg <- nrow(train_df %>%
                      filter(sentiment == "negative")) / nrow(train_df)

# Create bag of words for training

vocabulary <- unique(names(freq_word))
n_vocabulary <- length(vocabulary)

# Split the full_freq_df data by sentiment class 
# Each dataframe contains all words of the vocabulary as attributes

freq_df_pos <- full_freq_df %>%
  filter(sentiment == "positive") %>%
  select(1:195) 

freq_df_pos <- rbind(freq_df_pos , count = colSums(freq_df_pos))

freq_df_neg <- full_freq_df %>%
  filter(sentiment == "negative") %>%
  select(1:195) 

freq_df_neg <-  rbind(freq_df_neg , count = colSums(freq_df_neg))

# Calculate the count of all words in all documents of each class (including duplicates) 

total_count_pos <- rowSums(freq_df_pos[1001 , ])
total_count_neg <- rowSums(freq_df_neg[1002 , ])

# Calculate conditional probabilities for each term for both classes of sentiment
# Incorporate Laplace smoothing of +1

condprobs_neg <- sapply(freq_df_neg[1002 , ] , function(x) { (x + 1) / (total_count_neg + n_vocabulary) })
condprobs_pos <- sapply(freq_df_pos[1001 , ] , function(x) { (x + 1) / (total_count_pos + n_vocabulary) })

names(condprobs_neg) <- vocabulary
names(condprobs_pos) <- vocabulary

## Stage 2 ##

# Use the model on unseen test documents and predict the sentiment class

# Preprocess the test set documents in the same manner as the training documents
# Make a copy of the test set first

test_df2 <- test_df

test_corpus <- Corpus(VectorSource(test_df$review))
test_corpus <- tm_map(test_corpus , tolower)
test_corpus <- tm_map(test_corpus , removeNumbers)
test_corpus <- tm_map(test_corpus , removeWords , stopwords(kind = "english"))
test_corpus <- tm_map(test_corpus , removeWords , c("one" , "two" , "movie" , "movies", "film" , "films" , "get" , "make"))
test_corpus <- tm_map(test_corpus , removePunctuation)
test_corpus <- tm_map(test_corpus , stripWhitespace)
test_corpus <- tm_map(test_corpus , stemDocument , language = "english")

# Convert corpus back to dataframe (note that the only variable is now the review)

test_df <- data.frame(review = sapply(test_corpus , as.character) , stringsAsFactors = FALSE)

term_list <- lapply(test_df , function(x) { strsplit(x , split = " ") })
term_list <- term_list$review[1:1966]

# Now, we need to match up the terms with their corresponding aposterior probabilities
# for the target class and calculate the class probabilities for both sentiments

# Create vectors to store probabilities of sentiment being positive or negative

pred_pos <- vector("numeric")

for(i in term_list) { 
  inserts <- sapply(i, function(n) { n <- condprobs_pos[n] } ) 
  pred <- sum(log(inserts) , na.rm = T) + apriori_pos
  pred_pos[length(pred_pos)+1] <- pred
}

pred_neg <- vector("numeric")

for(i in term_list) { 
  inserts <- sapply(i, function(n) { n <- condprobs_neg[n] } ) 
  pred <- sum(unlist(log(inserts)) , na.rm = T) + apriori_neg
  pred_neg[length(pred_neg)+1] <- pred
}

preds_df <- data.frame(pred_neg , pred_pos)
names(preds_df) <- c("negative" , "positive")

# Bind the conditional probabilities to the test dataframe and also the winning classes and
# actual sentiment classes

winners <- colnames(preds_df)[max.col(preds_df)]
results_df <- cbind(test_df , preds_df , predicted = winners , actual = test_df2$sentiment)

# Obtain model accuracy

conf_matrix <- table(winners , test_df2$sentiment)
accuracy <- (conf_matrix[1,1] + conf_matrix[2,2]) / nrow(test_df2)

print(paste("Model accuracy: " , round(accuracy , 3) * 100 , "%"))

File: NB_sentiment_analysis_IMDB.py
# IMDB Movie Review Sentiment Analysis


# Import modules necessary for task
# Note that the "from __future__ import division" statement allows true division to occur when calculating probabilities

from __future__ import division, print_function
import string
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import nltk
import sklearn


# Create a function for converting the vector of sentiment labels to a Boolean factor with two levels
# True represents a positive review while False represents a negative review


def convert_to_boolean(sentiment_string):
    if sentiment_string == "negative":
        return False
    else:
        return True


## Stage 1 ##

# Read in the training and test documents provided with the assignment submission
# Assign the total number of documents in the training set to train_nb_documents

train_df = pd.read_csv("full_train_sentiment.csv")
train_nb_documents = train_df.shape[0]

# The sentiment labels need to be factorised
# This is done by converting the labels to Boolean levels

train_df["sentiment"] = train_df["sentiment"].apply(lambda x:convert_to_boolean(x))

test_df = pd.read_csv("sentiment_test.csv")
test_nb_documents = test_df.shape[0]

test_df["sentiment"] = test_df["sentiment"].apply(lambda x:convert_to_boolean(x))


## Stage 2 ##

# Some text preprocessing is required in order to prepare the documents for model building

# Convert all characters to lower case
train_df["review"] = train_df["review"].apply(lambda x: x.lower())

# Remove numbers
train_df["review"] = train_df["review"].apply(lambda review_string: ''.join([character for character in review_string if character not in "0123456789"]))

# Remove punctuation
train_df["review"] = train_df["review"].apply(lambda review_string: ''.join([character for character in review_string if character not in string.punctuation]))

# Remove stopwords
from nltk.corpus import stopwords
nltk.download("stopwords")

stopwords_english = set(stopwords.words("english"))
train_df["review"] = train_df["review"].apply(lambda review_string: ' '.join([word for word in review_string.split() if word not in stopwords_english]))

# Remove additional words which carry no real sentiment determining power
filtered_words = set(["one" , "two" , "movie" , "movies", "film" , "films" , "get" , "make"])
train_df["review"] = train_df["review"].apply(lambda review_string: ' '.join([word for word in review_string.split() if word not in filtered_words]))

# Carry out word stemming and add stemmed reviews as an additional column in the training dataframe
from nltk.stem.snowball import SnowballStemmer
stemmer = SnowballStemmer("english")

train_df["stemmed"] = train_df["review"].apply(lambda review_string: ' '.join([stemmer.stem(word) for word in review_string.split()]))

# The documents are much cleaner after this text preprocessing
# Inspect the first document after stemming

print(train_df.ix[0])

# # Create a document-term matrix which gives word frequency counts for each document
# # Notice how large the dimensions are (2001 rows of 30754 attributes)
from sklearn.feature_extraction.text import CountVectorizer
countvec = CountVectorizer()

# Build a pandas dataframe containing documents as rows and words as columns
train_dtm = pd.DataFrame(countvec.fit_transform(train_df["stemmed"]).toarray(), columns = countvec.get_feature_names())

# # Remove sparse terms to reduce this high dimensionality
# # Dimensions reduced to 2001 rows and 196 attributes
countvec2 = CountVectorizer(min_df = 0.2)
train_dtm = pd.DataFrame(countvec2.fit_transform(train_df["stemmed"]).toarray(), columns = countvec2.get_feature_names())
print(train_dtm.shape) 

# What are the word frequencies across all training documents?
# The vocabulary (bag of words) is the 196 terms in the freq_word series
freq_word = train_dtm.sum()
freq_word.sort_values(inplace = True, ascending = False)
print(freq_word)

n_vocabulary = len(freq_word)

# Add sentiment labels to the document-term matrix so we can calculate class apriori probabilities
train_dtm["sentiment"] = train_df["sentiment"]

apriori_pos = train_df["sentiment"].value_counts()[True] / train_nb_documents
apriori_neg = train_df["sentiment"].value_counts()[False] / train_nb_documents

# Split the training data by sentiment class
# Each subset contains all words of the vocabulary as attributes
freq_df_pos = train_dtm.loc[train_dtm["sentiment"] == True]
del(freq_df_pos["sentiment"])

freq_df_neg = train_dtm.loc[train_dtm["sentiment"] == False]
del(freq_df_neg["sentiment"])

# Calculate the total count of all words in all documents of each class (including duplicates)
total_count_pos = freq_df_pos.sum().sum()
total_count_neg = freq_df_neg.sum().sum()

freq_df_pos_count = freq_df_pos.sum()
freq_df_neg_count = freq_df_neg.sum()

# Calculate conditional probabilities for each term for both classes of sentiment
# Incorporate Laplace smoothing of +1

condprobs_neg = freq_df_neg_count.apply(lambda x: (x + 1) / (total_count_neg + n_vocabulary))
condprobs_pos = freq_df_pos_count.apply(lambda x: (x + 1) / (total_count_pos + n_vocabulary))


## Stage 3 ##

# Use the model on unseen test documents and predict the sentiment class

# Preprocess the test set documents in the same manner as the training documents

test_df["review"] = test_df["review"].apply(lambda x: x.lower() )
test_df["review"] = test_df["review"].apply(lambda review_string: ''.join([character for character in review_string if character not in "0123456789"]))
test_df["review"] = test_df["review"].apply(lambda review_string: ''.join([character for character in review_string if character not in string.punctuation]))
test_df["review"] = test_df["review"].apply(lambda review_string: ' '.join([word for word in review_string.split() if word not in stopwords_english]))
test_df["review"] = test_df["review"].apply(lambda review_string: ' '.join([word for word in review_string.split() if word not in filtered_words]))
print(test_df)

# Get conditional probabilities for each document in the test set being positively reviewed
pred_pos = test_df["review"].apply(lambda review_string: np.log(apriori_pos) + sum([np.log(condprobs_pos.get(word,1)) for word in review_string.split()]))

# Then calculate conditional probabilities for each document being negatively reviewed
pred_neg = test_df["review"].apply(lambda review_string: np.log(apriori_neg) + sum([ np.log(condprobs_neg.get(word,1)) for word in review_string.split()]))

# Display the results in a dataframe
results_df = pd.DataFrame()
results_df["P(positive)"] = pred_pos
results_df["P(negative)"] = pred_neg

# Add the predicted classes to the dataframe
# Remember that False represents a negative review while True represents a positive review
results_df["prediction"] = results_df["P(positive)"] > results_df["P(negative)"]

# Then add the actual classes of the test documents to the results dataframe
results_df["actual"] = test_df["sentiment"]

# Finally we calculate a model accuracy
from sklearn.metrics import confusion_matrix
conf_matrix = confusion_matrix(results_df["actual"], results_df["prediction"])
accuracy = (conf_matrix[0,0] + conf_matrix[1,1]) / test_nb_documents

print("Model accuracy:", round(accuracy , 3) * 100 , "%")

